iml-agent/src/action_plugins/check_ha.rs

```diff
@@ -4,90 +4,137 @@
+                    g.find_all("primitive")
+                        .map(move |p| create(p, Some(name.clone())))
+                })
+                .chain(
+                    elem.find_all("primitive")
+                        .map(|p| create(p, None)),
+                )
+                .collect(),
+        ),
     }
 }
 
-#[derive(serde::Deserialize, serde::Serialize, Clone, Debug)]
-pub struct AgentInfo {
-    pub agent: ResourceAgentType,
-    pub group: Option<String>,
-    pub id: String,
-    pub args: HashMap<String, String>,
+pub async fn check_ha(_: ()) -> Result<Vec<ResourceAgentInfo>, ImlAgentError> {
+    cmd_output_success("cibadmin", vec!["--query", "--xpath", "//resources"])
+        .await
+        .and_then(|output| process_resources(output.stdout.as_slice()))
 }
 
-impl AgentInfo {
-    pub fn create(elem: &Element, group: Option<String>) -> Self {
-        AgentInfo {
-            agent: ResourceAgentType::new(
-                elem.get_attr("class").unwrap_or("").to_string(),
-                elem.get_attr("provider").map(|s| s.to_string()),
-                elem.get_attr("type").unwrap_or("").to_string(),
-            ),
-            group,
-            id: elem.get_attr("id").unwrap_or("").to_string(),
-            args: match elem.find("instance_attributes") {
-                None => HashMap::new(),
-                Some(e) => e
-                    .find_all("nvpair")
-                    .map(|nv| {
-                        (
-                            nv.get_attr("name").unwrap_or("").to_string(),
-                            nv.get_attr("value").unwrap_or("").to_string(),
-                        )
-                    })
-                    .collect(),
-            },
-        }
-    }
-}
+#[cfg(test)]
+mod tests {
+    use super::process_resources;
 
-pub fn check_ha(_: ()) -> impl Future<Output = Result<Vec<AgentInfo>, ImlAgentError>> {
-    cmd_output_success("cibadmin", vec!["--query", "--xpath", "//resources"])
-        .and_then(|output| {
-            // This cannot split between map/map_err because Element does not implement Send
-            match Element::from_reader(output.stdout.as_slice()) {
-                Err(err) => future::err(ImlAgentError::XmlError(err)),
-                Ok(elem) => future::ok(
-                    elem.find_all("group")
-                        .flat_map(|g| {
-                            let name = g.get_attr("id").unwrap_or("").to_string();
-                            g.find_all("primitive")
-                                .map(move |p| AgentInfo::create(p, Some(name.clone())))
-                        })
-                        .chain(
-                            elem.find_all("primitive")
-                                .map(|p| AgentInfo::create(p, None)),
-                        )
-                        .collect(),
-                ),
-            }
-        })
-        .err_into()
+    #[test]
+    fn test_good() {
+        let testxml = r#"<resources>
```

Comment:
Possible to put in a separate file?

BTW, it seems missing an XML declaration - `<?xml version="1.0" ... ?>`

Reply:
It's not missing the declaration.  This is testing the output of `cibadmin --query --xpath //resources`